pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -370,12 +472,79 @@ func (b *Base) Objects(rsrc interface{}, rsrclabels map[string]string, observed,
 	}
 	var err error
 	for k, v := range configs {
-		resources, err = addConfigMapItem(master, getConfigName(master, k), labels, v, resources)
+		resources, err = addConfigMapItem(master, getResourceName(master, k), labels, v, resources)
 		if err != nil {
 			return nil, err
 		}
 	}
 
+	// User has reset the CR after failure. Remove upgrade failed status.
+	if master.Spec.Image == master.Status.ImageToUse && master.Status.Meta.IsConditionTrue(upgradeFailed) {
+		master.Status.ClearCondition(upgradeFailed, upgradeResetMessage, upgradeResetMessage)
+	}
+
+	if !imageToUseNeedsUpdate(master) {
+		// Set ImageToUse if upgrade has not failed
+		if !master.Status.Meta.IsConditionTrue(upgradeFailed) {
+			master.Status.ImageToUse = master.Spec.Image
+			master.Status.UserInterfaceImageToUse = master.Spec.UserInterfaceImage
+		}
+		return resources, nil
+	}
+
+	var upgradeResources []reconciler.Object
+	if master.Status.UpgradeJobVersion == "" {
+		// Add a new upgrade job, and set ready status to false
+		var err error
+		upgradeResources, err = prepareNewPreUpgradeJobResources(master, rsrclabels)
+		if err != nil {
+			return nil, err
+		}
+		addUpgradeComponentNotReady(master)
+		return append(resources, upgradeResources...), nil
+	}
+
+	item := k8s.GetItem(observed, &batchv1.Job{}, getUpgradeJobName(master), "default")
+	var upgradeJob *batchv1.Job
+	if item != nil {
+		upgradeJob = item.(*batchv1.Job)
+	}
+
+	if upgradeJob == nil {
+		// Job disappeared, add a new job and set ready status to false
+		var err error
+		upgradeResources, err = prepareNewPreUpgradeJobResources(master, rsrclabels)
+		if err != nil {
+			return nil, err
+		}
+		addUpgradeComponentNotReady(master)
+		return append(resources, upgradeResources...), nil
+	}
+
+	if upgradeJob.Status.Succeeded > 0 {
+		// Pre upgrade job has succeeded.
+		master.Status.ImageToUse = master.Spec.Image
+		master.Status.UserInterfaceImageToUse = master.Spec.UserInterfaceImage
+		master.Status.UpgradeJobVersion = ""
+		addUpgradeComponentNotReady(master)
+	} else if upgradeJob.Status.Failed >= upgradeFailureLimit {
+		// Pre upgrade job has failed
+		master.Status.UpgradeJobVersion = ""
+		master.Status.Meta.SetCondition(upgradeFailed, upgradeJobFailedMessage, upgradeJobFailedMessage)
+	} else {
+		// Pre upgrade job has already been initialized, but has not finished
+		var err error
+		upgradeResources, err = getPreUpgradeJobResources(master, rsrclabels)
+		if err != nil{
+			return nil, err
+		}
+		addUpgradeComponentNotReady(master)
+	}
+
+	if upgradeResources != nil {
```

Comment:
You can save this if by moving the `resources = append(resources, upgradeResources...)` inside the last `else` block.